### PR TITLE
Refactor rabbitmq_user provider (mpolenchuk)

### PR DESCRIPTION
**Pull request by wyardley**
This is an attempt to rework #460 by @mpolenchuk (keeping original attribution).
Could use careful review, rebasing was pretty hairy, and not sure I got everything.

As of now, failing at least one test.

### CODE REVIEW

**Comment by wyardley** [comment]
@ekohl @hunner I think this is ready for review, and would very much appreciate a review on this one if at all possible. In particular, I think the tests could use some improvement, especially for the change password situation.

However, at a functional level, I think this is working(ish), and an improvement.

File: spec/unit/puppet/provider/rabbitmq_user/rabbitmqctl_spec.rb [hunk 1]
```diff
@@ -3,213 +3,174 @@
...
+{ok,{user,<<"rmq_x">>,[],rabbit_auth_backend_internal,
+          {internal_user,<<"rmq_x">>,
+                         <<193,81,62,182,129,135,196,89,148,87,227,48,86,2,154,
+                           192,52,119,214,177>>,
+                         []}}}
+EOT
+      end
+
+      it do
+        provider.check_password('secret')
```
Comment (wyardley) [inline spec/unit/puppet/provider/rabbitmq_user/rabbitmqctl_spec.rb]: I don't think this is doing anything useful, but need to come up with a better test here.

**Comment by wyardley** [comment]
@mpolenchuk - don't know if you've got time to review, but let me know if this looks sane. I'm still trying to come up with a better test case for the password change bit.

File: lib/puppet/type/rabbitmq_user.rb [hunk 2]
```diff
@@ -40,20 +40,12 @@
...
     def change_to_s(_current, _desired)
       'password has been changed'
     end
-
-    def should_to_s(_newvalue = @should)
-      '<new password>'
-    end
   end
 
   newproperty(:admin) do
```
Comment (ekohl) [inline lib/puppet/type/rabbitmq_user.rb]: You can use ```newproperty(:admin, boolean: true, parent: Parent::Property::Boolean)``` and remove the body (except ```desc```). Then you can use true booleans in the provider. You might need a ```require 'puppet/property/boolean``` but I'm not sure about that.

Comment (wyardley) [reply lib/puppet/type/rabbitmq_user.rb]: I think this is a good idea, but I feel like it should be implemented separately, and then we can review all the type parameters at once, and adjust any that we can switch to true booleans / integers / etc in one 'breaking' changeset.

**Comment by wyardley** [comment]
Removing bug label since #147 was fixed via #629

File: lib/puppet/type/rabbitmq_user.rb (hunk 2 above)
Comment (ekohl) [reply lib/puppet/type/rabbitmq_user.rb]: That's fair

File: spec/unit/puppet/provider/rabbitmq_user/rabbitmqctl_spec.rb [hunk 3]
```diff
@@ -3,220 +3,174 @@
...
-one [administrator]
-foo []
-icinga  [monitoring]
-kitchen []
-kitchen2        [abc, def, ghi]
-EOT
-    provider.expects(:rabbitmqctl).with('set_user_tags', 'foo', %w[tag1 tag2])
-    provider.tags = %w[tag1 tag2]
-  end
+  describe '#check_password' do
```
Comment (wyardley) [inline spec/unit/puppet/provider/rabbitmq_user/rabbitmqctl_spec.rb]: Is this doing anything useful? Or should this (syncing password) be checked in the type spec instead, or by calling another method vs. the check_password one?

**Comment by wyardley** [comment]
> You could consider expanding the user acceptance test with purging

@ekohl added a test of ensuring absent on the user resource, which I think accomplishes the same thing, and also switched to the shared example now that #632 is merged.

**Comment by wyardley** [comment]
@ekohl: the shared example seems to get executed after the other elements in the block, which is causing an ordering issue.
For example
```
    it_behaves_like 'an idempotent resource'

    it 'does not have the user' do
      shell('rabbitmqctl list_users -q') do |r|
        expect(r.stdout).not_to match(%r{dan\s+})
      end
    end
```
results in
```
  destroy user resource
    does not have the user (FAILED - 1)
    behaves like an idempotent resource
localhost $ scp /var/folders/0z/yrnglm6n1l5_q19dv1yy0w3c0000gn/T/beaker20170913-21543-xcrm0o centos-7-x64:/tmp/apply_manifest.pp.Kz0FfQ {:ignore => }
      applies with no errors
localhost $ scp /var/folders/0z/yrnglm6n1l5_q19dv1yy0w3c0000gn/T/beaker20170913-21543-ez5qfe centos-7-x64:/tmp/apply_manifest.pp.BUFI7K {:ignore => }
      applies a second time without changes
```

It's failing because the test for the user being gone runs _before_ the shared example runs.

File: lib/puppet/type/rabbitmq_user.rb (hunk 2 above)
Comment (wyardley) [reply lib/puppet/type/rabbitmq_user.rb]: It seems like the existing code is (in some cases) as simple as:
```
  newparam(:force) do
    defaultto(:false)
    newvalues(:true, :false)
  end
```
which actually seems simpler than the suggested code in https://docs.puppet.com/puppet/5.1/custom_types.html, and without that extra `require`. Also, the puppet boolean type allows `'true'` `yes`, and various other things, which seems kind of regressive.

**Pull request merged**